src/libcore/iter/traits/iterator.rs

```diff
@@ -2373,29 +2375,35 @@ pub trait Iterator {
     /// assert_eq!(right, [2, 4]);
     /// ```
     #[stable(feature = "rust1", since = "1.0.0")]
-    fn unzip<A, B, FromA, FromB>(self) -> (FromA, FromB) where
+    fn unzip<A, B, FromA, FromB>(self) -> (FromA, FromB)
+    where
         FromA: Default + Extend<A>,
         FromB: Default + Extend<B>,
-        Self: Sized + Iterator<Item=(A, B)>,
+        Self: Sized + Iterator<Item = (A, B)>,
     {
-        fn extend<'a, A, B>(
-            ts: &'a mut impl Extend<A>,
-            us: &'a mut impl Extend<B>,
-        ) -> impl FnMut((A, B)) + 'a {
-            move |(t, u)| {
-                ts.extend(Some(t));
-                us.extend(Some(u));
+        let mut lhs = FromA::default();
+        let mut rhs = FromB::default();
+
+        #[inline]
+        fn extend_rhs<'a, A, B, T: Extend<B>>(rhs: &'a mut T) -> impl FnMut((A, B)) -> A + 'a {
+            move |(a, b)| {
+                rhs.extend(Some(b));
+                a
             }
         }
 
-        let mut ts: FromA = Default::default();
-        let mut us: FromB = Default::default();
+        lhs.extend((&mut self).map(extend_rhs(&mut rhs)));
+
+        // lhs.extend may not have fully consumed the iterator
+        rhs.extend(&mut self);
```

Comment:
We can't feed the `(A, B)` pair to `rhs` -- this will have to `map` to the second element alone.

Reply:
Oh, duh. That's what I get for not testing.

Reply:
Fixed